src/Elastic.Apm/Helpers/DbConnectionStringParser.cs

```diff
@@ -41,214 +46,376 @@ internal Destination ExtractDestination(string dbConnectionString, out bool wasF
+					{ "Port" , ParsePortValue },
+					{ "Address" , ParseAddressKeyWithNestedStructureValue },
+				};
+			}
+
+			internal void ParseFlatKeyValuePairs(string flatKeyValuePairs)
+			{
+				foreach (var keyValue in flatKeyValuePairs.Split(FlatKeyValuePairsSeparator))
+					ParseKeyValue(keyValue);
+			}
+
+			private void ParseKeyValue(string keyValue)
 			{
-				if (string.IsNullOrWhiteSpace(keyValueString)) continue;
+				if (string.IsNullOrWhiteSpace(keyValue)) return;
 
-				var keyValueArray = keyValueString.Split(KeyValueSeparator);
-				if (keyValueArray.Length != 2)
+				var keyValueSplit = keyValue.Split(new[] { KeyValueSeparator }, 2);
+				if (keyValueSplit.Length < 2)
 				{
-					_logger.Trace()?.Log("Encountered invalid key-value pair - skipping it."
-						+ " keyValueString: `{KeyValueString}'. keyValueArray: {KeyValueArray}. dbConnectionString: {DbConnectionString}."
-						, keyValueString, keyValueArray, dbConnectionString);
-					continue;
+					_logger.Trace()?.Log("Encountered key-value pair without value - skipping it."
+						+ " keyValue: `{KeyValueString}'. keyValueSplit: {KeyValueSplit}."
+						, keyValue, keyValueSplit);
+					return;
 				}
 
-				if (! KeyToPropertySetter.TryGetValue(keyValueArray[0].Trim(), out var propSetter)) continue;
+				var key = keyValueSplit[0].Trim();
+				var value = keyValueSplit[1].Trim();
 
-				if (result == null) result = new Destination();
+				// Skip unknown keys
+				if (!_keyToPropertySetter.TryGetValue(key, out var valueParser))
+				{
+					if (!HasNestedStructure(value)) return;
+
+					// Unless value for an unknown key has nested structure
+					// then we want to go into the value to see if we find known keys
+					ParseNestedStructure(value);
+					return;
+				}
 
 				try
 				{
-					propSetter(keyValueArray[1].Trim(), result);
+					valueParser(keyValueSplit[1].Trim());
+				}
+				catch (FormatException)
+				{
+					throw;
 				}
 				catch (Exception ex)
 				{
-					_logger.Trace()?.LogException(ex, "Encountered invalid value for a known key"
-						+ " - considering the whole connection string as invalid and returning null."
-						+ " keyValueString: `{KeyValueString}'. keyValueArray: {KeyValueArray}. dbConnectionString: {DbConnectionString}."
-						, keyValueString, keyValueArray, dbConnectionString);
-					return null;
+					throw new FormatException(
+						"Encountered invalid value for a known key."
+						+ $" keyValueSplit: {keyValueSplit}."
+						+ $" keyValue: `{keyValue}'."
+						, ex);
 				}
 			}
 
-			return result;
-		}
-
-		private static void ParseServerValue(string valueToParseArg, Destination destination)
-		{
-			var valueToParse = TrimDiscardable(valueToParseArg);
-
-			if (valueToParse.StartsWith(SqlServerLocalDbPrefix, StringComparison.OrdinalIgnoreCase)
-				|| valueToParse.StartsWith(SqlServerExpressUserInstancePrefix, StringComparison.OrdinalIgnoreCase))
+			private static bool HasNestedStructure(string value)
```

Comment:
nit: should be expression bodied member (Can't add suggestion to this one - I guess due to having added and removed lines here)